bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java

```diff
@@ -197,12 +197,14 @@ public void run() {
 			try {
 				listener = new ServerSocket(port);
 
-				socket = listener.accept();
-				outStr = new DataOutputStream(socket.getOutputStream());
-				while (socket.getInputStream()
-					.read() != -1) {
-					testRunSession = new TestRunSessionAndPort(launch, project);
-					outStr.writeInt(testRunSession.getPort());
+				while (true) {
```

Comment:
So the only way out of this loop is throwing an exception? That seems like poor flow control.

Reply:
The loop spends the idle time in the call to `accept()`, which blocks until it receives an incoming connection attempt. The only way (that I have found) to interrupt it is by closing the socket from another thread, which causes `accept()` to throw a `SocketException`. If you know a better way, I'm all ears!

Reply:
In the current code, it waits at the accept, then enters the read loop until the socket is closed and then exits the thread. In you change, it never, exits the thread. It just loop back up again. There is no control mechanism to tell this thread it is done and should exit. This makes this class generate thread which are stuck waiting on a server socket.

Reply:
Sorry, I may not have been clear. When I said "the only way to interrupt [`accept()`] is by closing the [server] socket from another thread", I should have perhaps noted that this mechanism was already in place. This code should not leave any threads behind stuck waiting on a server socket.

Zombie threads were still a potential problem that needed to be handled even in the old code, because it would still wait at the `accept()` until the first connection attempt was made - a zombie thread could still arise if that first connection attempt never came.

The fix was to call `close()` on the server socket at termination time:

https://github.com/bndtools/bnd/blob/5905abe02ef9c1687ca01192aa0c0d59b407dbd7/bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java#L219

https://github.com/bndtools/bnd/blob/5905abe02ef9c1687ca01192aa0c0d59b407dbd7/bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java#L245-L246

This in turn causes the `accept()` call in the background thread to throw a `SocketException`, which exits the loop and allows the thread to die naturally.

The same mechanism still works in the new implementation. You can verify that this is the case by running a test - when the app-under-test exits, you should see this log message in your log:

https://github.com/bndtools/bnd/blob/5905abe02ef9c1687ca01192aa0c0d59b407dbd7/bndtools.core/src/bndtools/launch/OSGiJUnitLaunchDelegate.java#L207-L208

Reply:
So this only applies to rerunIDE mode for tester.junit-platform?

Reply:
Yes.

It also occurred to me if that if the loop's control flow is not immediately obvious to someone with your experience then a little commentary might be in order. I'll add some comments.